### PR TITLE
Pyspec framework

### PR DESCRIPTION
**Pull request by lightclient**
This is the initial framework for an eth1 pyspec.

Ethereum 1.0 lacks a comprehensive and accessible specification. This creates several pain points, but the most prevalent is that intricate core changes must spend a [significant amount of time](https://eips.ethereum.org/EIPS/eip-1559) describing the current state of Ethereum, before even defining the proposed changes.

The [Ethereum 2.0 spec](https://github.com/ethereum/eth2.0-specs) has been a great example of "spec driven development". Eventually, we'd like to merge the two specs into a uniform, all encompassing specification. Until then, we've taken the liberty to experiment with a couple slightly different approaches than the eth2 specs:

* **_generate documentation from code_** -- the eth2 spec is a collection of markdown files that can be transpiled into an executable python project. My understanding is that this was a natural evolution of the spec, which started as mark down and slowly became more explicit, until the point where if the code snippets were concatenated, they could actually run and generate tests. Since we're starting from scratch, we can make an unbiased decision. While we agree that a "narrative" spec is desirable, we are confident that [Sphinx](https://www.sphinx-doc.org/en/master/) will give us complete control over the look and flow of the spec. It will require more effort than markdown, but we believe it will drastically simplify other portions of the toolchain (e.g. type checking and linting), make the raw spec more readable and maintainable, and allow us to use more tools that already exist instead of writing our own.
* **_non-forking code_** -- the eth2 spec generally understands the concept of "forks" and handles them accordingly. We would like the eth1 spec to completely ignore the concept of forks and instead allow it to be "the latest version of ethereum". We believe this will lead to a simpler spec and simpler proposals for core changes. Forks are an unavoidable part of Ethereum though, so we'd like to support them via a release mechanism. For example, `1.x.x` would be Frontier, `2.x.x` would be Homestead, etc. We plan to write a handler program that runs specific releases up until a fork block and then shuttle the current state of the chain into the next release.

So, with that said, we are currently working on implementing the Frontier release of Ethereum. This is very much a work-in-progress still and everything is up for discussion. We feel the sooner this work is in the open, the sooner we can get feedback from the community and potentially find additional contributors.

As of this PR, we've successfully:
* generated the mainnet genesis state root
* passed a load of RLP and trie tests
* passed a simple `add` + `sstore` test

We'll generally try to avoid writing our own tests wherever possible and instead rely on the `ethereum/tests` format. We plan to support the `t8n` interface so that the spec can generate test cases. We'd also like to eventually sync mainnet, but that may require some work arounds.

We're excited to hear people's thoughts and feedback :)

### CODE REVIEW

**Comment by timbeiko** [comment]
This is great work, huge props for taking it on :clap: 

One thought re: versioning: IMO it may make sense to keep all pre-merge versions of the spec `1.x` and use the merge as the start of the `2.0` version of the spec (so Frontier is 1.0, Homestead 1.1, etc.). This would align better with the "Eth 2.0" nomenclature. Even if we want to deprecate it in favor of execution (eth1) and consensus (eth2), it still seems coherent to have the merge be v2.0 of the execution layer. 

We can keep incrementing it long-term when we have significant changes to how the entire protocol works (e.g. state expiry) vs. "feature additions" in network upgrades such as Istanbul, Berlin, London, etc.

**Comment by lightclient** [comment]
> One thought re: versioning: IMO it may make sense to keep all pre-merge versions of the spec 1.x and use the merge as the start of the 2.0 version of the spec (so Frontier is 1.0, Homestead 1.1, etc.)

The downside of this is that this would incompatible with SemVer, which requires minor (`1.x.0`) changes be backwards compatible and major (`x.0.0`) be reserved for backwards incompatible changes.

File: README.md
Comment (pipermerriam) [inline README.md:28]: A small paragraph or sentence describing what this is seems appropriate here.

> The Consensus Specification is a python implementation of the .....  that prioritizes readability and simplicity.  It is accompanied by both narrative and API level documentation of the various components written in restructured text and rendered using Sphinx....

File: README.md
Comment (pipermerriam) [inline README.md:51]: Supporting pypy is noble but I would suggest a level of pragmatism if you run into things that this makes complex.  It is probably *ok* for this spec to have a very specific version and implementation of python for which it is considered "correct".

File: README.md
Comment (pipermerriam) [inline README.md:77]: we have a convention in our python repositories of having a `MAKE` command that will do these things for you: [`make lint-roll`](https://github.com/ethereum/ddht/blob/142911d134ff839f3f79ff8fe9e45d3fe5a58cd0/Makefile#L29-L40)

File: README.md
Comment (pipermerriam) [inline README.md:90]: Maybe "Accessible" instead of "Usable"?  Simple?  Low Barrier to Entry?

File: b2t.sh
Comment (pipermerriam) [inline b2t.sh:8]: *nitpick* 

expanding these flags to their longform is good self-documenting script code.

File: doc/_templates/autoapi/python/class.rst
Comment (pipermerriam) [inline doc/_templates/autoapi/python/class.rst:32]: can this file be documented/commented in some meaningful way?  Was it copied and modified from something existing?  Maybe inlining comments with examples of what any of the complex secitons or non-standard parts are intended to render?  this may be of little value so feel free to disregard.

File: main.py
Comment (pipermerriam) [inline main.py:46]: We have had good success with using `logging` over `print` which has the nice trade off of a little bit more boilerplate code which yields the ability to have different granularity about what is displayed when the *thing* is run.

https://github.com/ethereum/ddht/blob/142911d134ff839f3f79ff8fe9e45d3fe5a58cd0/ddht/logging.py#L59-L77

File: src/eth1spec/base_types.py
Comment (pipermerriam) [inline src/eth1spec/base_types.py:549]: Any reason this doesn't use [`typing.NewType`](https://docs.python.org/3/library/typing.html#newtype)?

File: src/eth1spec/eth_types.py
Comment (pipermerriam) [inline src/eth1spec/eth_types.py:80]: `Root` type appears to allow arbitrary length values (since it is an alias for `Bytes`).  Should this and the others be `Hash32`?

File: src/eth1spec/evm/__init__.py
Comment (pipermerriam) [inline src/eth1spec/evm/__init__.py:48]: Was any consideration given to using a `bytearray` here since it would give you a mutable data structure.  This would mirror the `stack` value which also looks to use a mutable data structure (but I haven't gotten far enough yet to know whether the mutable nature is actually used).

File: src/eth1spec/evm/instructions.py
Comment (pipermerriam) [inline src/eth1spec/evm/instructions.py:44]: I'm assuming it is intentional that this leaves out the overflow wrap-around.

File: src/eth1spec/base_types.py
Comment (pipermerriam) [inline src/eth1spec/base_types.py:316]: Returning this exception type rather than raising it is confusing to me.  Any specific reason for this?  If it was simply raised at the points where things aren't implemented then you would get clean stacktraces showing you where these are coming from.  As it stands now it will be hard to determine the origin of these errors.

File: src/eth1spec/spec.py
Comment (pipermerriam) [inline src/eth1spec/spec.py:178]: *nitpick*

Slightly more verbose (and thus maybe more readable) by removing the return statement inlining.

```python
...
gas_remaining = block_gas_limit - gas_available
post_root = trie.root(receipts_y)

return (gas_remaining, post_root, state)
```

**Review by pipermerriam — COMMENTED** [review]
3/4 of the way through and nervious that a commit is going to invalidate all the comments I've made so far.

**Comment by timbeiko** [comment]
> > One thought re: versioning: IMO it may make sense to keep all pre-merge versions of the spec 1.x and use the merge as the start of the 2.0 version of the spec (so Frontier is 1.0, Homestead 1.1, etc.)
> 
> The downside of this is that this would incompatible with [SemVer](https://semver.org/), which requires minor (`1.x.0`) changes be backwards compatible and major (`x.0.0`) be reserved for backwards incompatible changes.

That's fair, OTOH most projects who start with SemVer end up not strictly adhering to it 😅 I personally prefer "Ethereum consistency" over "Standards consistency", but it's not a very strong opinion.

File: src/eth1spec/trie.py
Comment (pipermerriam) [inline src/eth1spec/trie.py:116]: Just typing it out so that we can see whether it looks more readable using shifting and masking.  (not sure if this is actual equivalent logic)

```python
if is_even:
    nibbles_list[i] = key[i // 2] << 4
else:
    nibbles_list[i] = key[i // 2] & 0x0f
```

Code comments that just say "the high four bits" and "the low four bits" might be the most effective readability improvement here.

**Review by pipermerriam — APPROVED** [review]
Solid beginning.  I'm :+1: on getting this merged once it's been open for a few days to give people time to find it and :eyes:

File: b2t.sh
Comment (lightclient) [reply b2t.sh:8]: Unfortunately i) set doesn't have longform flags and ii) I didn't document _why_ I put those flags there so now I've forgotten! :(

File: src/eth1spec/evm/__init__.py
Comment (lightclient) [reply src/eth1spec/evm/__init__.py:48]: You're right, this should be a `bytearray`.

File: src/eth1spec/evm/instructions.py
Comment (lightclient) [reply src/eth1spec/evm/instructions.py:44]: Nope, this should've used `wrapping_add`. I didn't realize we were throwing a `ValueError` on overflow.

File: b2t.sh
Comment (pipermerriam) [reply b2t.sh:8]: https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html

See the `-o`:

```
set -o errexit -o nounset -o functrace
```

Though I'm not really sure this is objectively better now that I wrote it out....

File: src/eth1spec/spec.py
Comment (lightclient) [reply src/eth1spec/spec.py:333]: I agree with this, but I think for simplicity we'll keep it as-is until we get around to setting up a non-consensus module things like this.

File: README.md
Comment (SamWilsn) [reply README.md:51]: I think, as a mostly pure python project, it should be pretty easy to maintain compatibility, and I'd like to at least attempt to until we hit something that makes it difficult.

Ideally, we want to this project to sync mainnet, and PyPy might help speed that up. If that fails miserably, we can maybe fall back on Cython.

File: README.md
Comment (SamWilsn) [reply README.md:77]: So far I've managed to avoid using `make` for anything ;)

`tox` does the checks, but as long as the actual build doesn't rely on `make`, I guess I'd be okay with a script to simplify auto-formatting.

File: b2t.sh
Comment (SamWilsn) [reply b2t.sh:8]: Might specifically be `bash` syntax too. Other shells might only support the short options.

File: doc/_templates/autoapi/python/class.rst
Comment (SamWilsn) [reply doc/_templates/autoapi/python/class.rst:32]: All the templates are slightly modified versions from the upstream projects. I could at least link to the repository and commit I got them from though.

 - [ ] Add upstream links to templates.

File: src/eth1spec/base_types.py
Comment (SamWilsn) [reply src/eth1spec/base_types.py:549]: We haven't finished the type annotations yet :sweat_smile:

File: src/eth1spec/base_types.py
Comment (SamWilsn) [reply src/eth1spec/base_types.py:316]: `NotImplemented` isn't an exception (see https://docs.python.org/3/library/constants.html#NotImplemented).

This is just how these functions are supposed to work, apparently.

Comment (pipermerriam) [reply src/eth1spec/base_types.py:316]: I was reading these as: https://docs.python.org/3/library/exceptions.html#NotImplementedError

File: README.md
Comment (thomkeh) [reply README.md:53]: If the Python code is fully typed and basically only uses the standard library, then you could also compile it to C with [mypyc](https://github.com/mypyc/mypyc) for speed. It might not be as fast as pypy (though I haven't seen any benchmarks comparing them), but mypyc seems much more actively developed than pypy.

**Comment by pipermerriam** [comment]
I would nominate that this get merged soon so that additional changes can be done in smaller units and directly targeting this repository rather than the quilt fork.

**Comment by SamWilsn** [comment]
Agreed! None of us can actually do the merge though.

**Comment by pipermerriam** [comment]
Still seeing a stream of commits coming in.  You wanna give me a poke when you're ready with something stable.

**Comment by lightclient** [comment]
@pipermerriam we can keep it stable until Monday if someone wants to merge this. I'm guessing by then, we'll have some more PRs that we'll want to get merged.

**Comment by timbeiko** [comment]
@lightclient happy to merge this now if you think it's ready to go!

**Comment by lightclient** [comment]
It's ready! :)

**Pull request merged**
